docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,282 @@
+# Proposal:
+
+- Author(s):     [morgo](https://github.com/morgo)
+- Last updated:  March 09, 2021
+- Discussion at: N/A
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Test Design](#test-design)
+    * [Functional Tests](#functional-tests)
+    * [Scenario Tests](#scenario-tests)
+    * [Compatibility Tests](#compatibility-tests)
+    * [Benchmark Tests](#benchmark-tests)
+* [Impacts & Risks](#impacts--risks)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+This document was created to discuss the design of Dynamic Privileges. It is intended to be implemented in combination with Security Enhanced Mode, but there no interdependencies between the two features.
+
+## Motivation or Background
+
+MySQL 8.0 introduced the concept of “dynamic privileges” (see [WL#8131](https://dev.mysql.com/worklog/task/?id=8131)). The intention behind this functionality is that plugins can create new named privileges to suit their purposes, such as “Firewall Admin” or “Audit Admin” instead of requiring the `SUPER` privilege, which becomes overloaded and too coarse.
+
+Dynamic privileges are different from static privileges in the following ways:
+
+* The name of the privilege is “DYNAMIC”. A plugin registers it, and the server has no prior knowledge of its existence.
+* Privileges can only be global scoped.
+* Privileges are stored in the table `mysql.global_grants` and not `mysql.user`.
+* The `GRANT OPTION` is stored for each dynamic privilege (versus for the user as a whole).
+
+We have the same requirement for fine grained privileges in TiDB, so it makes sense to adopt a similar implementation of dynamic privileges. This document describes both the implementation of the framework for dynamic privileges and an initial set of dynamic privileges that are required to be implemented.
```

Comment:
How to add dynamic privileges without changing the TiDB code, recompile it, and redeploy the tidb-server?

Reply:
See https://github.com/pingcap/enterprise-plugin/pull/40 for an example adding `AUDIT_ADMIN`. There is no recompile, it is added to the plugin's `OnInit` method.

Reply:
TiDB Dashboard also needs the dynamic privilege feature so that different features can be authorized independently instead of a single SUPER privilege. SUPER privilege is also very danguerous.

Reply:
@breeswish if you know what they are, we can create a dynamic privilege such as `DASHBOARD_ADMIN` (follows same convention as `BACKUP_ADMIN`, `AUDIT_ADMIN` etc.)